contracts/CompoundFCM.sol

```diff
@@ -0,0 +1,337 @@
+    tickSpacing = _vamm.tickSpacing(); // retrieve tick spacing of the VAM
+
+    __Ownable_init();
+    __Pausable_init();
+    __UUPSUpgradeable_init();
+  }
+
+    // GETTERS FOR STORAGE SLOTS
+    // Not auto-generated by public variables in the storage contract, cos solidity doesn't support that for functions that implement an interface
+    /// @inheritdoc ICompoundFCM
+    function cToken() external view override returns (ICToken) {
+        return _ctoken;
+    }
+    /// @inheritdoc IFCM
+    function marginEngine() external view override returns (IMarginEngine) {
+        return _marginEngine;
+    }
+    /// @inheritdoc IFCM
+    function vamm() external view override returns (IVAMM) {
+        return _vamm;
+    }
+    /// @inheritdoc IFCM
+    function rateOracle() external view override returns (IRateOracle) {
+        return _rateOracle;
+    }
+
+  // To authorize the owner to upgrade the contract we implement _authorizeUpgrade with the onlyOwner modifier.
+  // ref: https://forum.openzeppelin.com/t/uups-proxies-tutorial-solidity-javascript/7786
+  function _authorizeUpgrade(address) internal override onlyOwner {}
+
+  event FullyCollateralisedSwap(
+    address indexed trader,
+    uint256 marginInScaledYieldBearingTokens,
+    int256 fixedTokenBalance,
+    int256 variableTokenBalance
+  );
+
+  function getTraderWithYieldBearingAssets(
+        address trader
+    ) external override view returns (TraderWithYieldBearingAssets.Info memory traderInfo) {
+      return traders[trader];
+    }
+
+
+  /// @notice Initiate a Fully Collateralised Fixed Taker Swap
+  /// @param notional Notional that cover by a fully collateralised fixed taker interest rate swap
+  /// @param sqrtPriceLimitX96 The binary fixed point math representation of the sqrtPriceLimit beyond which the fixed taker swap will not be executed with the VAMM
+  function initiateFullyCollateralisedFixedTakerSwap(uint256 notional, uint160 sqrtPriceLimitX96) external override {
+
+    require(notional!=0, "notional = 0");
+
+    /// todo: (URGENT) add support for approvals and recipient (similar to how it is implemented in the MarginEngine)
+
+    // initiate a swap
+    // the default tick range for a Position associated with the FCM is tickLower: -tickSpacing and tickUpper: tickSpacing
+    // isExternal is true since the state updates following a VAMM induced swap are done in the FCM (below)
+    IVAMM.SwapParams memory params = IVAMM.SwapParams({
+        recipient: address(this),
+        amountSpecified: notional.toInt256(),
+        sqrtPriceLimitX96: sqrtPriceLimitX96,
+        tickLower: -tickSpacing,
+        tickUpper: tickSpacing
+    });
+
+    (int256 fixedTokenDelta, int256 variableTokenDelta, uint256 cumulativeFeeIncurred, ,) = _vamm.swap(params);
+
+    require(variableTokenDelta <=0, "VT delta sign");
+
+    TraderWithYieldBearingAssets.Info storage trader = traders[msg.sender];
+
+    uint256 currentExchangeRate = _ctoken.exchangeRateCurrent();
```

Comment:
maybe worth adding a comment above this line explaining what exchangeRateCurrent means and what the math is for readability?

Reply:
Done